### PR TITLE
Add support for generated columns in SQLite3 adapter

### PR DESCRIPTION
**Pull request by fractaledmind**
### Motivation / Background

SQLite is a feature-rich database engine, and production usage is only growing. We need Rails' support to offer developers the range and scope of its features.

Both the MySQL and PostgreSQL adapters support virtual columns, and SQLite itself has support generated columns [since 2020](https://www.sqlite.org/releaselog/3_31_0.html).

### Detail

PR https://github.com/rails/rails/pull/41856 introduced virtual columns to the PostgreSQLAdapter. This is effectively a clone of that feature, with some necessary tweaks.

```ruby
create_table :users do |t|
  t.string :name
  t.virtual :name_upper, type: :string, as: 'UPPER(name)'
  t.virtual :name_lower, type: :string, as: 'LOWER(name)', stored: true
end
```

Firstly, SQLite supported both `STORED` and `VIRTUAL` generated columns, while PostgreSQL only supports `VIRTUAL`. Secondly, SQLite doesn't support `ALTER TABLE` commands with generated columns. 

In this PR, I am adding full support for the SQLite3Adapter by implementing:

`ActiveRecord::ConnectionAdapters::SQLite3::Column#virtual?`
`ActiveRecord::ConnectionAdapters::SQLite3Adapter#supports_virtual_columns?`

and altering:

`ActiveRecord::ConnectionAdapters::SQLite3::SchemaCreation#add_column_options`
`ActiveRecord::ConnectionAdapters::SQLite3::SchemaDefinitions#new_column_definition`
`ActiveRecord::ConnectionAdapters::SQLite3::SchemaDumper#prepare_column_options`
`ActiveRecord::ConnectionAdapters::SQLite3::SchemaStatements#new_column_from_field`
`ActiveRecord::ConnectionAdapters::SQLite3Adapter#table_structure`
`ActiveRecord::ConnectionAdapters::SQLite3Adapter#invalid_alter_table_type?`
`ActiveRecord::ConnectionAdapters::SQLite3Adapter#table_structure_with_collation`

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/sqlite3/column.rb [hunk 1]
```diff
@@ -1,15 +1,18 @@
 # frozen_string_literal: true
 
+require "active_support/core_ext/object/blank"
```
Comment (skipkayhil) [inline activerecord/lib/active_record/connection_adapters/sqlite3/column.rb:3]: `object/blank` is already required in all frameworks by default:
https://github.com/rails/rails/blob/162be9f2196a02c72b5b15944352a33b94e47b6e/activerecord/lib/active_record.rb#L27
https://github.com/rails/rails/blob/162be9f2196a02c72b5b15944352a33b94e47b6e/activesupport/lib/active_support/rails.rb#L14

**Review by skipkayhil — COMMENTED** [review]
Small comment, but generally looks good

File: activerecord/lib/active_record/connection_adapters/sqlite3/column.rb (hunk 1 above)
Comment (fractaledmind) [reply activerecord/lib/active_record/connection_adapters/sqlite3/column.rb:3]: Interesting. I copied this from the PostgreSQL adapter, so I will remove it from there as well.

File: activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb [hunk 2]
```diff
@@ -185,6 +193,16 @@ def quoted_scope(name = nil, type: nil)
             scope[:type] = type if type
             scope
           end
+
+          def extract_generated_type(field)
+            case field["hidden"]
+            when nil then nil
+            when 0 then nil
+            when 1 then nil
```
Comment (nvasilevski) [inline activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb:201]:
```suggestion

```

`case` should return `nil` for no match anyway so it feels unnecessary to declare these matches. Unless we want to be explicit about all possible values

**Review by nvasilevski — APPROVED** [review]
👍

File: activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb (hunk 2 above)
Comment (fractaledmind) [reply activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb:201]: Yeah, I defaulted to explicit to make code reviewer hopefully easier and clearer. Easier to remove code that is understood than add code that isn't there. 

So, I'll remove these matches now

**Comment by fractaledmind** [comment]
@zzak the fix for this flaky test is in main, can we rebuild this PR?

**Comment by fractaledmind** [comment]
@skipkayhil Addressed your one comment. Everything should be good to go now

File: activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb [hunk 3]
```diff
@@ -245,7 +245,7 @@ def test_insert_logged
           name = "foo"
 
           sqlite_version_query = ["SELECT sqlite_version(*)", "SCHEMA", []]
-          pragma_query = ["PRAGMA table_info(\"ex\")", "SCHEMA", []]
+          pragma_query = ["PRAGMA table_xinfo(\"ex\")", "SCHEMA", []]
```
Comment (yahonda) [inline activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:248]: Does SQLite version 3.30.1 or an older version that does not support generated columns have the `table_xinfo` pragma?

Comment (fractaledmind) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:248]: `PRAGMA table_xinfo` was added in version [3.26.0 (2018-12-01)](https://www.sqlite.org/changes.html#version_3_26_0). So, if the database supports generated columns (added in version [3.31.0 (2020-01-22)](https://www.sqlite.org/changes.html#version_3_31_0)), then it supports `table_xinfo`.

File: activerecord/lib/active_record/connection_adapters/sqlite3/column.rb [hunk 4]
```diff
@@ -20,6 +21,18 @@ def auto_incremented_by_db?
           auto_increment? || rowid
         end
 
+        def virtual?
+          @generated_type.present?
```
Comment (byroot) [inline activerecord/lib/active_record/connection_adapters/sqlite3/column.rb:25]:
```suggestion
          @generated_type
```

`.present?` isn't a "to_boolean" cast, it has fairly peculiar rules, and is rarely what you want.

Here we can just return `@generated_type`, or if you are adamant  about returning a boolean, you can use `!!@generated_type` or `!@generated_type.nil?`

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 5]
```diff
@@ -503,7 +511,8 @@ def has_default_function?(default_value, default)
         # SQLite has an additional restriction on the ALTER TABLE statement
         def invalid_alter_table_type?(type, options)
           type.to_sym == :primary_key || options[:primary_key] ||
-            options[:null] == false && options[:default].nil?
+            options[:null] == false && options[:default].nil? ||
+            (type.to_sym == :virtual && options[:stored])
```
Comment (byroot) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:515]: We should do the `to_sym` casting in `add_column`

**Review by byroot — APPROVED** [review]
Two minor nitpick. I also rebased your branch to replace the merge commit.

Given how small the changes are I'll apply them myself.

**Pull request merged**